### PR TITLE
RUN-5060 : window.navigate should ack nack same as new fin...Window

### PR DESCRIPTION
**Pull request by tomer-openfin**
#### Description of Change
Making sure win.navigate nacks.

#### Checklist
- [x] PR description included
- [x] `npm test` passes
- [x] PR title starts with the JIRA ticket [pull request process](https://github.com/openfin/Internal-Wiki/wiki/Pull-Request-Process)
- doc change already present in canary

### CODE REVIEW

File: src/browser/api/window.js [hunk 1]
```diff
@@ -1502,8 +1502,28 @@ Window.moveTo = function(identity, left, top) {
...
+            if (success) {
+                resolve();
+            } else {
+                reject();
+            }
+        };
+
+        let didFail = () => handleLoadFinished(false);
+        let didSucceed = () => handleLoadFinished(true);
+        browserWindow.webContents.on('did-fail-load', didFail);
```
Comment (datamadic) [inline src/browser/api/window.js]: Do we get the actual reason here (HTTP response code etc)? If so, lets not squash the reason but report it in the context from which it came.

File: src/browser/api/window.js [hunk 2]
```diff
@@ -1502,8 +1502,28 @@ Window.moveTo = function(identity, left, top) {
...
+            browserWindow.webContents.removeListener('did-fail-load', didFail);
+            browserWindow.webContents.removeListener('did-finish-load', didSucceed);
+            if (success) {
+                resolve();
+            } else {
+                reject();
+            }
+        };
+
+        let didFail = () => handleLoadFinished(false);
```
Comment (rdepena) [inline src/browser/api/window.js]: Hey @tomer-openfin Looks like we are loosing some fidelity here, [Electron will return](https://electronjs.org/docs/api/web-contents#event-did-fail-load) the [Chromium net error](https://cs.chromium.org/chromium/src/net/base/net_error_list.h), take a look at how these errors are [handled in window creation](https://github.com/HadoukenIO/core/blob/01727fe61bb91994243ebe1fe45ad6f258b215da/src/browser/api/window.js#L776) and let's make sure we format our errors in a similar fashion.

File: src/browser/api/window.js [hunk 3]
```diff
@@ -1502,18 +1502,60 @@ Window.moveTo = function(identity, left, top) {
 };
 
 Window.navigate = function(identity, url) {
-    let browserWindow = getElectronBrowserWindow(identity, 'navigate');
-    browserWindow.webContents.loadURL(url);
+    return new Promise((resolve, reject) => {
+        const browserWindow = getElectronBrowserWindow(identity);
+
+        if (!browserWindow) {
```
Comment (rdepena) [inline src/browser/api/window.js]: You can delete these checks, we have middleware to check [entity existence](https://github.com/HadoukenIO/core/blob/develop/src/browser/api_protocol/api_handlers/middleware_entity_existence.ts)

File: src/browser/api/window.js [hunk 4]
```diff
@@ -1502,18 +1502,60 @@ Window.moveTo = function(identity, left, top) {
...
+    });
 };
 
 Window.navigateBack = function(identity) {
-    let browserWindow = getElectronBrowserWindow(identity, 'navigate back');
-    browserWindow.webContents.goBack();
+    return new Promise((resolve, reject) => {
+        const browserWindow = getElectronBrowserWindow(identity);
+
+        if (!browserWindow) {
```
Comment (rdepena) [inline src/browser/api/window.js]: You can delete these checks, we have middleware to check [entity existence](https://github.com/HadoukenIO/core/blob/develop/src/browser/api_protocol/api_handlers/middleware_entity_existence.ts)

File: src/browser/api/window.js [hunk 5]
```diff
@@ -1502,18 +1502,60 @@ Window.moveTo = function(identity, left, top) {
...
+            const error = new Error(`Unknown window named '${identity.name}'`);
+            return reject(error);
+        }
+
+        if (!browserWindow.webContents.canGoBack()) {
+            const error = new Error(`Can't navigate back - history is empty`);
```
Comment (rdepena) [inline src/browser/api/window.js]: Let's change the error message here. Would change it to something a bit more generic: 

'Cannot navigate back'

**Comment by datamadic** [comment]
Is [this test](https://testing-dashboard.openfin.co/#/app/tests/5c588e8acb360141a7dfdae1/edit) an expected failure?

**Comment by datamadic** [comment]
Also, @tomer-openfin is this something that we can test for in the automated suite?

**Comment by tomer-openfin** [comment]
This [test](https://testing-dashboard.openfin.co/#/app/tests/5c588e8acb360141a7dfdae1/edit) now passes.
Also added a [test](https://testing-dashboard.openfin.co/#/app/tests/5cdc7f1bcb360141a7dfe3a4/edit) for the ack\nack functionality (would appreciate any feedback)

**Pull request merged**